### PR TITLE
djl-serving version 0.11 (new formula)

### PR DESCRIPTION
**Pull request by aksrajvanshi**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)? 
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb:17]:
```suggestion
    (bin/"djl-serving").write_env_script "#{libexec}/bin/djl-serving", Language::Java.overridable_java_home_env
```

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb:46]: Can you propose a more substantial test than checking for the existence of some files? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application.

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

**Review by carlocab — COMMENTED** [review]
Thanks for your PR, @aksrajvanshi.

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb]: This seems like a good test to run. Can you configure this to use a different port than `8080`? `free_port` will give you an available port. Look at other formulae that use it for examples; there are about a hundred of them.

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: I actually changed the test to something simpler and which still verifies that the program ran successfully

Comment (carlocab) [reply Formula/djl-serving.rb]: Yes, I can see the test you have now. What I mean is that this test is better.

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: Yes, but I realized that it isn't a good test since the port is configured in the code itself. Moreover, including a link was creating a lot of problems with the test.

Comment (carlocab) [reply Formula/djl-serving.rb]: CI was complaining about the line length. You could always declare the command in a separate line:
```ruby
cmd = "curl -k -X POST 'http://127.0.0.1:8080/models?url=https%3A%2F%2Fdjl-ai.s3.amazonaws.com%2Fresources%2Ftest-models%2Fmlp.tar.gz&initial_workers=1&model_name=mlp2'"
assert_includes shell_output(cmd), "registered."
```
You can even interpolate the long URL into `cmd` if `cmd` is too long.

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: Ok I 've made the changes. Thank you for the help.

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb:51]: Shouldn't you be forking a process rather than using `shell_output` for something it's not really designed for here?

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb]: The test already runs in a temporary directory. You can recover this path using `testpath`.

File: Formula/djl-serving.rb
Comment (aksrajvanshi) [reply Formula/djl-serving.rb:51]: yeah. I am having trouble with this. Also, I have another doubt. This script starts a process at port 8080. to run the curl command and connect to the script/server, we can't possibly use `free_port`? Am I right ? Or am I missing something here ?

File: Formula/djl-serving.rb
Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: The temporary directory I'm trying to create is to create a cache to download necessary engines.

Comment (carlocab) [reply Formula/djl-serving.rb]: So why can't you use `testpath` for that?

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb:49]:
```suggestion
    fork do
      with_env(DJL_CACH_DIR: testpath) do
        exec bin/"djl-serving"
      end
    end
```
Please have a look at the other formulae that use `free_port` or `fork`, as I've already suggested, so you have a better idea of how to do this properly.

Comment (carlocab) [reply Formula/djl-serving.rb:49]: If `with_env` doesn't work, you can always set it with
```ruby
ENV["DJL_CACHE_DIR"] = testpath
```

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb]:
```suggestion
      (testpath/"config.properties").write <<~EOS
        echo inference_address=http://127.0.0.1:6035
        echo management_address=http://127.0.0.1:6035
      EOS
      with_env(
        DJL_CACHE_DIR: testpath,
        SERVING_OPTS: "-Djna.debug_load=true -Djna.debug_load.jna=true"
      ) do
```
You can probably write `config.properties` outside the `fork do` block. Can we use `free_port` instead of `6035`?

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: Yes, I'm working on using the `free_port`. Sure, I'll take the config.properties out of the fork block. We're facing some problem with JNA library where they're not downloading during the build run, whereas it's running on my system.

Comment (carlocab) [reply Formula/djl-serving.rb]: Great. You can do something like
```ruby
port = free_port
(testpath/"config.properties").write <<~EOS
  echo inference_address=http://127.0.0.1:#{port}
  echo management_address=http://127.0.0.1:#{port}
EOS
```

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: Hello, I wanted to ask you about the permissions on the build. For my code to run/pass, I would need write access on the machine. Do we have such access when dealing with tests, `testpath` on the build ?

Comment (carlocab) [reply Formula/djl-serving.rb]: Yes, you have write access to `testpath` and its subdirectories. You almost certainly don't have write access elsewhere.

File: Formula/djl-serving.rb
Comment (SMillerDev) [inline Formula/djl-serving.rb:50]:
```suggestion
    (testpath/"config.properties").write <<~EOS
    inference_address=http://127.0.0.1:6000
    management_address=http://127.0.0.1:6000
EOS
```

File: Formula/djl-serving.rb
Comment (SMillerDev) [inline Formula/djl-serving.rb:46]: We don't really need to check if brew can install files. We have unittests for that.

File: Formula/djl-serving.rb
Comment (SMillerDev) [inline Formula/djl-serving.rb:60]:
```suggestion
    assert_match "something", shell_output("curl http://127.0.0.1:#{port}/ping")
```

File: Formula/djl-serving.rb
Comment (SMillerDev) [inline Formula/djl-serving.rb:61]: This should be satisfied by calling the port right?

File: Formula/djl-serving.rb
Comment (SMillerDev) [inline Formula/djl-serving.rb:62]: Anything that is started in a brew test will be automatically killed as well. This shouldn't be needed.

File: Formula/djl-serving.rb
Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: The ping command works fine and responds as `status:healthy` when I test it. It seems like the connection is getting refused in the build system. Can something be done to resolve this ?

Screen Shot 2021-03-23 at 11 21 04 AM

**Comment by SMillerDev** [comment]
It's failing to write a log to libexec on ARM, but the log should really be written to `#{var}/log` and to `#{testpath}/log` in the test.

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb:57]:
```suggestion
    assert_match "{\n  \"status\": \"Healthy\"\n}\n", shell_output("curl #{cmd}")
```
This is a bit easier to read and more consistent with other tests.

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb:51]: Why do we need this line? Is it because the test will attempt to write into `libexec` for logging otherwise?

Also, are you able to make sure that this writes logs into `#{var}/log` rather than into `libexec`?

**Review by carlocab — COMMENTED** [review]
Nice work here. I just have a few more questions.

File: Formula/djl-serving.rb
Comment (aksrajvanshi) [reply Formula/djl-serving.rb:51]: Yes, we have made sure the logs aren't written to `libexec`. I can check the effect of removing line 51.

Comment (carlocab) [reply Formula/djl-serving.rb]: What about during regular use of `djl-serving`, outside of the `test` block? Where does it write its logs?

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: By default, it writes the logs to `libexec/logs` folder. It depends on the `MODEL_SERVER_HOME` environment variable.

Comment (carlocab) [reply Formula/djl-serving.rb]: That's what I was asking earlier. It shouldn't be writing there. Instead, it should be writing into `#{var}/log`. We can wrap `djl-serving` in a wrapper that sets the `MODEL_SERVER_HOME` variable in a way that can be overridden by the user.

Just add it to the `env` argument in your `write_env_script` call.

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: So it'll be `(bin/"djl-serving", MODEL_SERVER_HOME).write_env_script "#{libexec}/bin/djl-serving", Language::Java.overridable_java_home_env`

Comment (carlocab) [reply Formula/djl-serving.rb]: It probably has to be something like
```ruby
env = { MODEL_SERVER_HOME: "${MODEL_SERVER_HOME:-#{var}/log}" }
env.merge!(Language::Java.overridable_java_home_env)
(bin/"djl-serving").write_env_script "#{libexec}/bin/djl-serving", env
```
Where does `djl-serving` store its logs relative to `MODEL_SERVER_HOME`? Does it store it in `MODEL_SERVER_HOME`, or does it use something like `${MODEL_SERVER_HOME}/logs`?

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: It's inside `${MODEL_SERVER_HOME}/logs`

Comment (carlocab) [reply Formula/djl-serving.rb]: Then the first line should probably just be
```ruby
env = { MODEL_SERVER_HOME: "${MODEL_SERVER_HOME:-#{var}}" }
```
Does that seem right to you, @SMillerDev? 

Or, I guess this might also work:
```ruby
env = { MODEL_SERVER_HOME: "${MODEL_SERVER_HOME:-#{var}/djl-serving}" }
```

Comment (aksrajvanshi) [reply Formula/djl-serving.rb]: The `env = { MODEL_SERVER_HOME: "${MODEL_SERVER_HOME:-#{var}}" }` seems to be the way in macos in general.

Comment (carlocab) [reply Formula/djl-serving.rb]: I guess we can go with that then.

File: Formula/djl-serving.rb
Comment (carlocab) [inline Formula/djl-serving.rb:57]:
```suggestion
    sleep 30
```
Let's make this longer. CI runs things in a VM which makes some things run oddly slow sometimes.

Comment (aksrajvanshi) [reply Formula/djl-serving.rb:57]: Do you want me to change the sleep time ? I think all the CI builds passed this time.

Comment (carlocab) [reply Formula/djl-serving.rb:57]: Yes. Sleep time issues usually only emerge in CI runs which test a lot of formulae (e.g. when testing `openjdk`), so we may not have had any problems here, but it might still be an issue later.

Squash your commit together when you make this change, please. Thanks!

**Comment by aksrajvanshi** [comment]
Do I need to do add anything for the branch merging ? Or we need multiple approvals for the merging to take place ?

**Comment by carlocab** [comment]
I was waiting to see if @SMillerDev had a comment on https://github.com/Homebrew/homebrew-core/pull/73649#discussion_r600116449, but it being marked as resolved might've prevented them from seeing it.

**Comment by aksrajvanshi** [comment]
> I was waiting to see if @SMillerDev had a comment on [#73649 (comment)](https://github.com/Homebrew/homebrew-core/pull/73649#discussion_r600116449), but it being marked as resolved might've prevented them from seeing it.

Ohh! I will un-resolve it then :) You can notify them too if you can.

**Review by SMillerDev — APPROVED** [review]

**Comment by carlocab** [comment]
Thanks for your work on this, @aksrajvanshi!

**Pull request closed without merging**
